src/turbomind/models/llama/LlamaLinear.h

```diff
@@ -34,29 +35,69 @@ class LlamaLinear {
                  Type                       type      = kGemm,
                  int*                       lora_mask = nullptr)
     {
+        float           alpha     = 1.0f;
+        float           beta      = 0.0f;
+        GemmS4F16::Type gemm_type = type == kFusedSiluFfn ? GemmS4F16::kFusedSiluFfn : GemmS4F16::kGemm;
+        if (lora_mask != nullptr && weight.lora.r > 0) {
+            FT_CHECK(type == kGemm);
+            // output = lora(x) * scale
+            // output = mask(output)
+            // output = x*W + output
+            cublas_wrapper_->Gemm(CUBLAS_OP_N,
+                                  CUBLAS_OP_N,
+                                  weight.lora.r,                                  // m
+                                  batch_size,                                     // n
+                                  weight.input_dims,                              // k
+                                  (const T*)weight.lora.a,                        // A
+                                  weight.lora.r,                                  // lda
+                                  input_data,                                     // B
+                                  weight.input_dims,                              // ldb
+                                  output_data + batch_size * weight.output_dims,  // C
+                                  weight.lora.r);                                 // ldc
+
+            cublas_wrapper_->Gemm(CUBLAS_OP_N,
+                                  CUBLAS_OP_N,
+                                  weight.output_dims,                             // m
+                                  batch_size,                                     // n
+                                  weight.lora.r,                                  // k
+                                  (const T*)weight.lora.b,                        // A
+                                  weight.output_dims,                             // lda
+                                  output_data + batch_size * weight.output_dims,  // B
+                                  weight.lora.r,                                  // ldb
+                                  output_data,                                    // C
+                                  weight.output_dims,                             // ldc
+                                  weight.lora.scale,                              // alpha
+                                  0.0f);                                          // beta
+
+            invokeMask(output_data, lora_mask, batch_size, weight.output_dims, stream_);
+
+            beta      = 1.0f;
+            gemm_type = weight.type == WeightType::kINT4 ? GemmS4F16::kFusedAdd : gemm_type;
```

Comment:
We can set the type to `kFusedAdd` here for both FP and INT4 gemm.

1. In `forwardFp`, set beta to 1.0 depending on the type
2. In `forwardInt4`, convert type to `GemmS4F16::Type` for INT4 gemm interface

So that all differences between FP/INT4 gemms are kept local.